### PR TITLE
[SPMD] Add apply_backward_optimization_barrier

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request adds a new API to xla_sharding.py called apply_backward_optimization_barrier where registers a full backward hook that apply an optimization barrier to the given module. This API will prevent the XLA compiler from fusing the module's backward pass with others. And It's useful to prevent gigantic buffers being allocated to synchronize the gradients.

It's also used in https://github.com/pytorch-tpu/transformers/pull/50.

Test Plan:
python test/spmd/test_xla_sharding.py -v -k test_backward_optimization_barrier

### CODE REVIEW

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 1]
```diff
@@ -640,3 +640,23 @@ def backward(ctx, grad_output):
 
 def xla_patched_nn_linear_forward(m, input):
   return XLAPatchedLinear.apply(input, m.weight, m.bias)
+
+
+def apply_backward_optimization_barrier(m: torch.nn.Module):
```
Comment (jonb377) [inline torch_xla/distributed/spmd/xla_sharding.py:645]: Does this need to be applied to each layer of the model? What's the expected entrypoint for e.g. Llama2?

Comment (jonb377) [reply torch_xla/distributed/spmd/xla_sharding.py:645]: Oh I missed your link to the transformers repo - thanks!

File: test/spmd/test_xla_sharding.py [hunk 2]
```diff
@@ -1047,6 +1047,22 @@ def test_from_cpu_shards_global_shape(self):
       with self.assertRaises(RuntimeError):
         from_cpu_shards(shards, op_sharding, torch.Size((1,)))
 
+  def test_backward_optimization_barrier(self):
+    model = self.SimpleLinear().to(xm.xla_device())
+    # The first layer won't have gradients in the hook. Not sure why.
+    xs.xla_sharding.apply_backward_optimization_barrier(model.fc2)
+
+    # optimizer.zero_grad()
```
Comment (alanwaketan) [inline test/spmd/test_xla_sharding.py:1055]: I should remove this. oops...

**Comment by alanwaketan** [comment]
Thanks Jon for approving the pull request.

**Review by yeounoh — APPROVED** [review]
LGTM thanks!

**Review by JackCaoG — APPROVED** [review]
